core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportClusterAllocationExplainAction.java

```diff
@@ -113,14 +113,26 @@ protected void masterOperation(final ClusterAllocationExplainRequest request, fi
                 // If looking for a replica, go through all the replica shards
                 List<ShardRouting> replicaShardRoutings = allocation.routingTable().shardRoutingTable(index, shard).replicaShards();
                 if (replicaShardRoutings.size() > 0) {
-                    // Pick the first replica at the very least
-                    foundShard = replicaShardRoutings.get(0);
-                    // In case there are multiple replicas where some are assigned and some aren't,
-                    // try to find one that is unassigned at least
-                    for (ShardRouting replica : replicaShardRoutings) {
-                        if (replica.unassigned()) {
-                            foundShard = replica;
-                            break;
+                    final String replicaNodeId = request.getCurrentNodeId();
+                    if (replicaNodeId != null) {
```

Comment:
This code already fails if a replica shard on the specified node cannot be found.  `foundShard` will be null so we will throw an exception below.